src/host/multiple_ledger.h

```diff
@@ -0,0 +1,815 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "consensus/ledger_enclave_types.h"
+#include "ds/logger.h"
+#include "ds/messaging.h"
+
+#include <cstdint>
+#include <cstdio>
+#include <filesystem>
+#include <limits>
+#include <list>
+#include <map>
+#include <string>
+#include <sys/types.h>
+#include <unistd.h>
+#include <vector>
+
+namespace fs = std::filesystem;
+
+namespace asynchost
+{
+  static constexpr size_t max_chunk_threshold_size =
+    std::numeric_limits<uint32_t>::max(); // 4GB
+  static constexpr size_t max_read_cache_size_default = 5;
+  static constexpr auto ledger_committed_suffix = ".committed";
+  static constexpr auto ledger_start_idx_delimiter = "_";
+
+  static inline bool is_ledger_file_compacted(const std::string& file_name)
+  {
+    auto pos = file_name.find(ledger_committed_suffix);
+    return !(pos == std::string::npos);
+  }
+
+  static inline size_t get_start_idx_from_file_name(
+    const std::string& file_name)
+  {
+    auto pos = file_name.find(ledger_start_idx_delimiter);
+    if (pos == std::string::npos)
+    {
+      throw std::logic_error(fmt::format(
+        "Ledger file name {} does not contain a start idx", file_name));
+    }
+
+    return std::stol(file_name.substr(pos + 1));
+  }
+
+  class LedgerFile
+  {
+  private:
+    using positions_offset_header_t = size_t;
+
+    static constexpr auto file_name_prefix = "ledger";
+
+    static constexpr size_t frame_header_size = sizeof(uint32_t);
+
+    const std::string dir;
+
+    size_t start_idx = 1;
+    size_t total_len = 0;
+    std::vector<uint32_t> positions;
+
+    // This uses C stdio instead of fstream because an fstream
+    // cannot be truncated.
+    FILE* file;
+    bool completed = false;
+    bool compacted = false;
+
+  public:
+    LedgerFile(const std::string& dir, size_t start_idx) :
+      dir(dir),
+      file(NULL),
+      start_idx(start_idx)
+    {
+      file = fopen(
+        (fs::path(dir) /
+         fs::path(fmt::format("{}_{}", file_name_prefix, start_idx)))
```

Comment:
Can we split this for readability? `file_name = ...; file_dir = ...; file = fopen(file_dir);`